### PR TITLE
Tiered List HOC/Pattern

### PR DESCRIPTION
**Pull request by pastelcyborg**
## Done

- Added new Tiered List macro which supports all variants from Figma
- Added example pages for each Tiered List variant, as well as a combined page
- Added new docs page for Tiered List, showcasing new Macro documentation
- (Drive-by) Add doc updates to separators clarifying usage
- (Drive-by) Add new `variants_all` scss file for combined usage

Fixes [WD-13041](https://warthogs.atlassian.net/browse/WD-13041)

## QA

- Open [docs page](https://vanilla-framework-5212.demos.haus/docs/patterns/tiered-list) and each new Tiered List example on the [examples page](https://vanilla-framework-5212.demos.haus/docs/examples)
- Review Tiered List pattern, Macro, and documentation

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

[WD-13041]: https://warthogs.atlassian.net/browse/WD-13041?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: scss/_patterns_split-list.scss [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@mixin vf-p-split-list {
+  * {
+    // border: 1px solid red;
+  }
+
+  .p-split-list-classes {
+    @extend .u-fixed-width;
+
+    .p-split-list__main-hr {
+      @extend .u-hide--small;
```
Comment (bartaz) [inline scss/_patterns_split-list.scss:12]: If we are writing CSS we should use borders, not hr elements.

File: scss/_patterns_split-list.scss [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    .p-split-list__main-row {
+      @extend .row;
+    }
+
+    .p-split-list__main-col {
+      &:first-child {
+        @extend .col;
```
Comment (bartaz) [inline scss/_patterns_split-list.scss:21]: Not sure what it does? `.col` is used in `row--x-x` shorthands, if we mean full width column we should make it more explicit with `.col-12`.

**Comment by bartaz** [comment]
In the CSS version currently we still seem to rely too much on HTML structure and having a lot of structural elements. All those custom rows/cols nested are something I'd like to avoid, because if we need to do that, there is almost no benefit of just using standard grid.

First of all we should get rid of `<hr>`s. If we are writing CSS we should use borders, not depend on `<hr>` existence. 
Then we should try to create somewhat semantic HTML structure rather than recreate a grid "rows" and "cols".

Somewhat like that (although I guess it may not cover all cases we need):

```
.p-split-list // extends row (50/50 or full width)
     .p-split-list__heading // extends column (size depends on layout)
         h2
     .p-split-list__list // extends column (size depends on layout)
         .p-split-list__list-item // extends a row (nested)
              .p-split-list__list-title // extends column in nested row
                   h3.p-heading--5 / h3.p-split-list__list-heading
              .p-split-list__list-content // extends column in nested row
                   p // not sure if we need paragraph there? can there be more than 1 paragraph of text there?
         .p-split-list__list-item
              .p-split-list__list-title
              .p-split-list__list-content
     .p-split-list__cta
```

File: templates/docs/examples/patterns/split-list.html [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+     'title': 'Rich portfolio of products',
+     'description': 'Resell the full range of Canonical’s portfolio of private and public cloud products: Openstack, Kubernetes, IoT, support, security and compliance for Ubuntu.'
+   }, {
+     'title': 'Sales enablement',
+     'description': 'Canonical can help train your field sales and presales teams and provide you with sales collateral and permission to use the Canonical and Ubuntu trademarks in your own materials.'
+   }]) }}
+
+<h6>full width, full width, 50-50 (Jinja macro)</h6>
+
+{{ split_list(variant=4,
```
Comment (bartaz) [inline templates/docs/examples/patterns/split-list.html:487]: Not sure, how different is this `split_list` version from the `_compact` one? Both take plain text arguments, right?

File: templates/docs/examples/patterns/split-list.html [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+   title='H2 - this can take up to two lines, but best to stick to one line.',
```
Comment (bartaz) [inline templates/docs/examples/patterns/split-list.html:488]: This is fine to assume it's always h2 for now and only take content, but it needs to support inline HTML (like `<code>` for example, or links).

File: templates/docs/examples/patterns/split-list.html [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+   list_items=[{
+     'title': 'Rich portfolio of products',
```
Comment (bartaz) [inline templates/docs/examples/patterns/split-list.html:490]: Same as title above, needs to support inline HTML.

File: templates/docs/examples/patterns/split-list.html [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+     'description': 'Resell the full range of Canonical’s portfolio of private and public cloud products: Openstack, Kubernetes, IoT, support, security and compliance for Ubuntu.'
```
Comment (bartaz) [inline templates/docs/examples/patterns/split-list.html:491]: We need a full HTML support here. This may be paragraphs, but also may be lists, images, etc (as per Lyubos latest designs).

Would it work for it to be multiline strings?

Can this content be defined elsewhere and referenced here (via some variables) potentially? (like when it will be taken from some kind of CMS or other external source).

File: templates/docs/examples/patterns/split-list.html (hunk 3 above)
Comment (bartaz) [inline templates/docs/examples/patterns/split-list.html:487]: Variant would need to be something more user friendly.

If there are two separate options, maybe there should be 2 props:
layout-large="50/50" | "100", etc
layout-medium="50/50" | "100", etc

with one of them being a default, so in most common case you don't have to define it at all?

**Review by bartaz — COMMENTED** [review]
Nice to see macros giving some value.

Some questions inline in the code.

File: templates/_macros/split-list.jinja [hunk 7]
```diff
@@ -7,7 +7,11 @@
 # title: split list title text
 # list_items: array of content objects for list
 #   [{ 'title': string, 'description': string }]
-{% macro split_list(variant=1, title='', list_items=[{}]) -%}
+{% macro split_list(variant=1, title='', subtitle='', list_items=[{}]) -%}
```
Comment (bartaz) [inline templates/_macros/split-list.jinja:10]: I guess it would make sense to call it `title` and `description`, similar to title/description inside the list?

File: templates/_macros/split-list.jinja
Comment (jmuzina) [inline templates/_macros/split-list.jinja:1]: When viewing the markup rendered by this macro in your documentation page, there is a lot of extra whitespace (probably generated by Jinja):

image

You may need to play a bit with the whitespace-stripping (`-%}` and `{%-`) to get it to be formatted nicely in embedded examples

File: templates/docs/examples/patterns/split-list/default.html
Comment (jmuzina) [inline templates/docs/examples/patterns/split-list/default.html:1]: All of the non-partial templates here should be included in a `examples/patterns/split-list/combined.html` file before final merging to reduce the number of Percy snapshots needed to test them

File: templates/docs/patterns/split-list/index.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+## Full-width w/ description
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/full-width-with-description/" class="js-example">
+View example of the matrix pattern
+</a></div>
+
+## Import
+
+To import and use this Macro in your project, follow the steps on the
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:129]: As far as I'm aware, just importing and using the macro will not also include the SCSS used by the macro - and this macro requires `vf-p-headings`.  

We should probably include instructions for doing that here, similar to the [other components](https://vanilla-framework-5212.demos.haus/docs/patterns/slider#import)

File: templates/_macros/split-list.jinja [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      {% endif %}
+    {% else %}
+      <div class="row">
+        <div class="col-12">
+          {{ title_content }}
+        </div>
+      </div>
+    {% endif %}
+
+    {% for number in range(1, 10) %}
```
Comment (jmuzina) [inline templates/_macros/split-list.jinja:48]: I wonder if we should alert the user in some way if they pass in more list items than are allowed - such as with a console error? Or should we simply rely on them to notice when they are missing content they expect to see?

File: templates/docs/patterns/split-list/index.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+[Macro installation guide](TODO).
```
Comment (bartaz) [inline templates/docs/patterns/split-list/index.md:130]: If there is no installation guide yet, let's not document it as such.

At this point it technically doesn't matter if it's a macro or not (if devs can't use it). We also don't really have a practice of documenting these macros, so need to figure that out.

I'd just suggest documenting it as a standard HTML for now, with examples to copy (internally using macros of course). And we can build on top of that when we have means to install and document those macros.

**Comment by bartaz** [comment]
I believe we talked about not calling it "Split list" (because of split list component). In Figma it's currently called "Key value", but I'm not a huge fan of that name either. I asked there (in Figma) in a comment.

File: templates/docs/patterns/split-list/index.md (hunk 10 above)
Comment (bartaz) [reply templates/docs/patterns/split-list/index.md:130]: Also, this link is 404, causing linkchecker to fail.

File: scss/standalone/patterns_split-list.scss [hunk 11]
```diff
@@ -0,0 +1,4 @@
+@import '../vanilla';
```
Comment (bartaz) [inline scss/standalone/patterns_split-list.scss:1]: We have to figure out how to get rid of those for patterns. 

Patterns by definition depend on other components, we would have to list them all here. There is no point.

This one is likely missing a bunch as well: grid, rule, CTA, buttons, etc…

File: side-navigation.yaml [hunk 12]
```diff
@@ -98,6 +98,8 @@
       url: /docs/patterns/segmented-control
     - title: Slider
       url: /docs/patterns/slider
+    - title: Split list
```
Comment (bartaz) [inline side-navigation.yaml:101]: Should be in new "Patterns" category (as per Hero PR)

File: templates/_macros/split-list.jinja
Comment (bartaz) [reply templates/_macros/split-list.jinja:1]: This may be tricky to get fully cleaned, but indeed a good use of `{%-    -%}` should help.

File: templates/_macros/split-list.jinja (hunk 9 above)
Comment (bartaz) [reply templates/_macros/split-list.jinja:48]: We need to verify with Lyubo if there is an upper limit.

I guess it doesn't hurt us too much if we increase it up to 50 or 100 (which should be way more than enough?)

**Comment by jmuzina** [comment]
No longer 404/500 errors on Percy tests as #5250 has been merged - Percy tests seem to be working properly

File: templates/_macros/split-list.jinja [hunk 13]
```diff
@@ -0,0 +1,103 @@
+# Params
+# is_description_full_width_on_desktop: whether description has its own row on
+#   desktop vs. 50/50 split between title/description
+# is_list_full_width_on_tablet: whether list title/description each have their
+#   own row on tablet vs. 50/50 split between title/description
```
Comment (jmuzina) [inline templates/_macros/split-list.jinja:5]: This may be personal preference, but it could be nice to also document the available slots here

File: templates/docs/patterns/split-list/index.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+View example of the matrix pattern
+</a></div>
+
+## 50/50 on tablet w/o description
+
+This variant does not contain a top-level description and its child list is
+presented with its titles side-by-side with its descriptions on tablet screen
+sizes.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/50-50-tablet-no-description/" class="js-example">
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:28]: Not a show-stopper, and not sure if this was a stylistic choice or not, but there are some places throughout these examples that could use a bit more whitespace control to get rid of extra newlines
image

File: templates/docs/patterns/split-list/index.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+This variant contains a top-level description. Its title and description are
+presented side-by-side on desktop screen sizes, and its child list is presented
+side-by-side on tablet screen sizes.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/50-50-with-description/" class="js-example">
+View example of the matrix pattern
+</a></div>
+
+## Full-width w/o description
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:51]: These "without <slot>" and "with <slot>" examples make me wonder if Hero should have something similar as well. 

It's rather verbose in the current structure of our docs (having a whole new example that is the same as the previous one but with/without one DOM node), but it's the best way of doing this right now.

File: templates/docs/patterns/split-list/index.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/full-width-no-description/" class="js-example">
+View example of the matrix pattern
+</a></div>
+
+## Full-width w/ description
+
+This variant contains a top-level description, and its title, description, and
+child list are presented full-width on desktop and tablet screen sizes
+respectively.
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:65]: This feels a bit wordy. Maybe something like:

```suggestion
This variant contains a top-level description. Its contents are presented full-width on desktop and tablet screen sizes.
```

File: templates/docs/patterns/split-list/index.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+View example of the matrix pattern
+</a></div>
+
+## 50/50 on tablet w/ description
+
+This variant contains a top-level description and its child list is presented
+with its titles side-by-side with its descriptions on tablet screen sizes.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/50-50-tablet-with-description/" class="js-example">
+View example of the matrix pattern
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:38]: Backup / fallback text for these examples needs to be updated to reflect split list pattern, not matrix pattern
```suggestion
View example of the split-list pattern
```

File: templates/docs/patterns/split-list/index.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+composition of your content, you can choose from a variety of split list
+layouts.
+
+## 50/50 on desktop w/ description
+
+This variant contains a top-level description which is presented side-by-side
+with its title on desktop screen sizes.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/50-50-desktop-with-description/" class="js-example">
+View example of the matrix pattern
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:19]:
```suggestion
View example of the split-list pattern
```

File: templates/docs/patterns/split-list/index.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+</a></div>
+
+## 50/50 w/ description
+
+This variant contains a top-level description. Its title and description are
+presented side-by-side on desktop screen sizes, and its child list is presented
+side-by-side on tablet screen sizes.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/50-50-with-description/" class="js-example">
+View example of the matrix pattern
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:48]:
```suggestion
View example of the split-list pattern
```

File: templates/docs/patterns/split-list/index.md [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+</a></div>
+
+## Full-width w/o description
+
+This variant does not contain a top-level description, and both its title and
+child list are presented full-width on desktop and tablet screen sizes
+respectively.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/full-width-no-description/" class="js-example">
+View example of the matrix pattern
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:58]:
```suggestion
View example of the split-list pattern
```

File: templates/docs/patterns/split-list/index.md [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+</a></div>
+
+## Full-width w/ description
+
+This variant contains a top-level description, and its title, description, and
+child list are presented full-width on desktop and tablet screen sizes
+respectively.
+
+<div class="embedded-example"><a href="/docs/examples/patterns/split-list/full-width-with-description/" class="js-example">
+View example of the matrix pattern
```
Comment (jmuzina) [inline templates/docs/patterns/split-list/index.md:68]:
```suggestion
View example of the split-list pattern
```

**Review by jmuzina — CHANGES_REQUESTED** [review]

**Comment by lyubomir-popov** [comment]
@pastelcyborg Looks great, just 3 comments:

- name change as aghreed to "tiered list"
- h2 headings need to be wrapped in p-section--shallow. We [discussed](https://github.com/canonical/vanilla-framework/issues/5260)with @bartaz the option of altering the default margin on h1/h2, but that's probably for a separate pr.
- all horizontal rules should be muted. We use the regular ones only at the top of sections, above the h2s. Bonus points if you add this to the documentation.

**Comment by lyubomir-popov** [comment]
- these two do the same as `u-fixed-width`, which uses one less div:

image

- p-section--shallow needed to wrap all the tiered list items:

image

- I don't think we need one around the description para at the top, but if you feel strongly about it, that's fine.

image

**Comment by lyubomir-popov** [comment]
Looks great! thanks @pastelcyborg, I've added the +1 label.

File: templates/docs/patterns/tiered-list/index.md [hunk 22]
```diff
@@ -0,0 +1,75 @@
+---
+wrapper_template: '_layouts/docs.html'
+context:
+  title: Tiered list | Patterns
+---
+
+The tiered list pattern is used to easily implement a list of paired titles and
+descriptions, complete with a top-level title and description. An optional CTA
+block is placed at the bottom of the content as well. Depending on the size and
```
Comment (jmuzina) [inline templates/docs/patterns/tiered-list/index.md:9]: Nit: link to CTA block component documentation here perhaps?

File: releases.yml [hunk 23]
```diff
@@ -12,6 +12,10 @@
       url: /docs/patterns/navigation#full-width-dropdowns
       status: New
       notes: We've introduced a new full-width dropdown option to the top navigation component.
+    - component: Tiered list pattern
```
Comment (jmuzina) [inline releases.yml:15]:
```suggestion
    - component: Tiered list
```
[Hero PR](https://github.com/canonical/vanilla-framework/pull/5248) does not use "pattern" here, maybe we should come to group decision on how to name these in release notes? I don't think they need "pattern"

See [Hero 4.15 release notes](https://vanilla-framework-5248.demos.haus/docs/whats-new)

File: templates/docs/examples/patterns/tiered-list/full-width-no-description.html [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    <h2>H2 - up to two lines; ideally one.</h2>
+  {%- endif -%}
+
+  {%- if slot == 'list_item_title_1' -%}
+    <h5>Rich portfolio of products</h5>
+  {%- endif -%}
+
+  {%- if slot == 'list_item_description_1' -%}
+    <p>Resell the full range of Canonical’s portfolio of private and public
+    cloud products: Openstack, Kubernetes, IoT, support, security and
```
Comment (jmuzina) [inline templates/docs/examples/patterns/tiered-list/full-width-no-description.html:20]: Nit: this and other instances of "Openstack" should be "OpenStack"

File: templates/docs/examples/patterns/tiered-list/full-width-no-description.html [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    transformation and expand your business through marketing support and
+    alignment with a trusted brand</p>
+  {%- endif -%}
+
+  {%- if slot == 'list_item_title_5' -%}
+    <h5>Technical enablement</h5>
+  {%- endif -%}
+
+  {%- if slot == 'list_item_description_5' -%}
+    <p>We provide in depth technical training for your customer engineering
```
Comment (jmuzina) [inline templates/docs/examples/patterns/tiered-list/full-width-no-description.html:58]: Nit: I think maybe these instances of "in depth" should be "in-depth"
https://www.beedictionary.com/common-errors/indepth_vs_in_depth

**Review by jmuzina — APPROVED** [review]
A couple nit-picks here and there on word choice and grammar, but I think this is ready for approval once those are sorted :)

**Review by jmuzina — DISMISSED** [review]

**Comment by bartaz** [comment]
Let's merge heros first, as this may need some rebasing and updating after.

**Review by jmuzina — APPROVED** [review]

File: templates/docs/patterns/tiered-list/index.md [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+  title: Tiered list | Patterns
+---
+
+The tiered list pattern is used to easily implement a list of paired titles and
+descriptions underneath a top-level title and description. An optional [CTA
+block](/docs/patterns/cta-block) is placed at the bottom of the content as well.
+Depending on the size and composition of your content, you can choose from a
+variety of tiered list layouts:
+
+- [50/50 on desktop w/ description](#5050-on-desktop-w-description)
```
Comment (bartaz) [inline templates/docs/patterns/tiered-list/index.md:13]: I find the `w/` `w/o` notation hard to read. I don't know how common it is or what are the rules of using it or not, but I feel like it's not really necessary.

It's especially confusing when combined with legitimate use of / in `50/50` or in our example titles:

image

"Tiered list / 50/50 on desktop w/ description" - how do you read that?

I'm also pretty sure any screen reader will fail on that.

So, please just use "with" and "without" instead everywhere: links, headings in the docs, titles of examples, etc.

**Review by jmuzina — APPROVED** [review]

**Review by bartaz — APPROVED** [review]
Ship it 🚢
